asserts/pool.go

```diff
@@ -19,6 +19,538 @@
+// point ToResolve can be called to get them organized in groupings
+// ready for fetching. Fetched assertions can then be provided with
+// Add or AddBatch. Because these can have prerequisites calling
+// ToResolve and fetching needs to be repeated until ToResolve's
+// result is empty.  Between any two ToResolve invocations but after
+// any Add or AddBatch AddUnresolved/AddToUpdate can also be used
+// again.
+//
+//                      V
+//                      |
+//        /-> AddUnresolved, AddToUpdate
+//        |             |
+//        |             V
+//        |------> ToResolve -> empty? done
+//        |             |
+//        |             V
+//        \ __________ Add
+//
+//  * TODO: AddToUpdate is not implemented yet
+//  * TODO: document Add*Error, CommitTo
+type Pool struct {
+	groundDB RODatabase
+
+	numbering map[string]uint16
+	groupings *internal.Groupings
+
+	unresolved    map[string]*unresolvedRec
+	prerequisites map[string]*unresolvedRec
+
+	bs     Backstore
+	groups map[uint16]*groupRec
+
+	curPhase poolPhase
+}
+
+// NewPool creates a new Pool, groundDB is used to resolve trusted and
+// predefined assertions and to provide the current revision for
+// assertions to update and their prerequisites. Up to n groups can be
+// used to organize the assertions.
+func NewPool(groundDB RODatabase, n int) *Pool {
+	groupings, err := internal.NewGroupings(n)
+	if err != nil {
+		panic(fmt.Sprintf("NewPool: %v", err))
+	}
+	return &Pool{
+		groundDB:      groundDB,
+		numbering:     make(map[string]uint16),
+		groupings:     groupings,
+		unresolved:    make(map[string]*unresolvedRec),
+		prerequisites: make(map[string]*unresolvedRec),
+		bs:            NewMemoryBackstore(),
+		groups:        make(map[uint16]*groupRec),
+	}
+}
+
+func (p *Pool) groupNum(group string) (gnum uint16, err error) {
+	if gnum, ok := p.numbering[group]; ok {
+		return gnum, nil
+	}
+	gnum = uint16(len(p.numbering))
+	if err = p.groupings.WithinRange(gnum); err != nil {
+		return 0, err
+	}
+	p.numbering[group] = gnum
+	return gnum, nil
+}
+
+func (p *Pool) ensureGroup(group string) (gnum uint16, err error) {
+	gnum, err = p.groupNum(group)
+	if err != nil {
+		return 0, err
+	}
+	var gRec *groupRec
+	if gRec = p.groups[gnum]; gRec == nil {
+		gRec = new(groupRec)
+		p.groups[gnum] = gRec
+	}
+	return gnum, nil
+}
+
+// Singleton returns a grouping containing only the given group.
+// It is useful mainly for tests and to drive Add are AddBatch when the
+// server is pushing assertions (instead of the usual pull scenario).
+func (p *Pool) Singleton(group string) (Grouping, error) {
+	gnum, err := p.ensureGroup(group)
+	if err != nil {
+		return Grouping(""), nil
+	}
+
+	var grouping internal.Grouping
+	p.groupings.AddTo(&grouping, gnum)
+	return Grouping(p.groupings.Label(&grouping)), nil
+}
+
+// An unresolvedRec tracks a single unresolved assertion until is is resolved
+// or there is an error doing so. grouping grows as members all the groups
```

Comment:
```suggestion
// An unresolvedRec tracks a single unresolved assertion until it is resolved
// or there is an error doing so. Grouping grows as members all the groups
```

Reply:
Suggested two typo fixes but second sentence sounds weird, is it missing something?

Reply:
grouping refers to the field in the struct, that's why it was lower case but the sentence can probably use some tweaks:

The field "grouping" will grow to contain as members all the groups requiring this assertion.

Is that clearer?

Reply:
Thank you, I got it now. Perhaps this would be even clearer: 

The field "grouping" will grow to contain all the groups requiring this assertion.

Or is it going too far?

Reply:
changed